File: shared/java/top/fpsmaster/ui/custom/ComponentsManager.java
package top.fpsmaster.ui.custom;

import net.minecraft.client.gui.ScaledResolution;
import org.lwjgl.opengl.GL11;
import top.fpsmaster.features.impl.InterfaceModule;
import top.fpsmaster.features.impl.interfaces.ClientSettings;
import top.fpsmaster.ui.custom.impl.*;
import top.fpsmaster.utils.Utility;
import top.fpsmaster.utils.render.Render2DUtils;

import java.util.ArrayList;

public class ComponentsManager {
    // List to hold all components
    public final ArrayList<Component> components = new ArrayList<>();

    // Variable to track drag lock state
    public String dragLock = "";

    // Initialize all components
    public void init() {
        components.add(new FPSDisplayComponent());
        components.add(new ArmorDisplayComponent());
        components.add(new MusicComponent());
        components.add(new ScoreboardComponent());
        components.add(new PotionDisplayComponent());
        components.add(new CPSDisplayComponent());
        components.add(new KeystrokesComponent());
        components.add(new ReachDisplayComponent());
        components.add(new ComboDisplayComponent());
        components.add(new LyricsComponent());
        components.add(new InventoryDisplayComponent());
        components.add(new TargetHUDComponent());
        components.add(new PlayerDisplayComponent());
        components.add(new PingDisplayComponent());
        components.add(new CoordsDisplayComponent());
        components.add(new ModsListComponent());
        components.add(new MiniMapComponent());
        components.add(new SprintComponent());
        components.add(new ItemCountDisplayComponent());
    }

    // Get a component by its class type
    public Component getComponent(Class<? extends InterfaceModule> clazz) {
        return components.stream()
                .filter(component -> component.mod.getClass() == clazz)
                .findFirst()
                .orElse(null);
    }

    // Draw all components on the screen
    public void draw(int mouseX, int mouseY) {
        GL11.glPushMatrix();

        // Adjust mouse coordinates if fixed scale is enabled
        if (ClientSettings.fixedScale.getValue()) {
            ScaledResolution sr = new ScaledResolution(Utility.mc);
            int scaleFactor = ClientSettings.fixedScale.getValue() ? sr.getScaleFactor() : 2;
            float guiWidth = sr.getScaledWidth() / 2f * scaleFactor;
            float guiHeight = sr.getScaledHeight() / 2f * scaleFactor;

            mouseX = mouseX * scaleFactor / 2;
            mouseY = mouseY * scaleFactor / 2;

            Render2DUtils.fixScale();
        }

        // Draw all components that should be displayed
        int finalMouseX = mouseX;
        int finalMouseY = mouseY;
        components.forEach(component -> {
            if (component.shouldDisplay()) {
                try {
                    component.display(finalMouseX, finalMouseY);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });

        GL11.glPopMatrix();
    }
}

File: shared/java/top/fpsmaster/ui/custom/ComponentsManager.kt
package top.fpsmaster.ui.custom

import net.minecraft.client.gui.ScaledResolution
import org.lwjgl.opengl.GL11
import top.fpsmaster.features.impl.InterfaceModule
import top.fpsmaster.features.impl.interfaces.ClientSettings
import top.fpsmaster.ui.custom.impl.*
import top.fpsmaster.utils.Utility
import top.fpsmaster.utils.render.Render2DUtils
import java.util.function.Consumer

class ComponentsManager {
    var components = ArrayList<Component>()
    var dragLock = ""

    fun init() {
        components.add(FPSDisplayComponent())
        components.add(ArmorDisplayComponent())
        components.add(MusicComponent())
        components.add(ScoreboardComponent())
        components.add(PotionDisplayComponent())
        components.add(CPSDisplayComponent())
        components.add(KeystrokesComponent())
        components.add(ReachDisplayComponent())
        components.add(ComboDisplayComponent())
        components.add(LyricsComponent())
        components.add(InventoryDisplayComponent())
        components.add(TargetHUDComponent())
        components.add(PlayerDisplayComponent())
        components.add(PingDisplayComponent())
        components.add(CoordsDisplayComponent())
        components.add(ModsListComponent())
        components.add(MiniMapComponent())
    }

    fun getComponent(clazz: Class<out InterfaceModule>): Component {
        return components.stream().filter { component: Component -> component.mod.javaClass == clazz }
            .findFirst().orElse(null)
    }

    fun draw(mouseX: Int, mouseY: Int) {
        GL11.glPushMatrix();
        var mouseX = mouseX
        var mouseY = mouseY
        if (ClientSettings.fixedScale.value){
            val sr = ScaledResolution(Utility.mc)
            var scaleFactor: Int = if (ClientSettings.Companion.fixedScale.value) {
                sr.scaleFactor;
            } else {
                2;
            }
            val guiWidth = sr.scaledWidth / 2f * scaleFactor
            val guiHeight = sr.scaledHeight / 2f * scaleFactor
            mouseX = mouseX * scaleFactor / 2
            mouseY = mouseY * scaleFactor / 2

            Render2DUtils.fixScale()
        }
        components.forEach(Consumer { component: Component ->
            if (component.shouldDisplay()) component.display(
                mouseX,
                mouseY
            )
        })
        GL11.glPopMatrix();
    }
}
